Windows/Arm64: Use 8.1 atomic instructions if they are available

kunalspathak opened the pull request:
Currently, the base version for armarch is v8.0. However, starting v8.1, there are faster `atomic` instructions that takes less instructions and do not need full barrier. See example: https://godbolt.org/z/x1o9Ybh66. This restricts the machine that supports the CPU feature from using these faster instructions. As part of this PR, added a flag `g_arm64_atomics_present` that is set at the startup when we get the CPU capability info. Below APIs are rewritten such that it checks this flag and use the faster instructions if atomic support is available on the machine:

- InterlockedCompareExchange
- InterlockedCompareExchangeAcquire
- InterlockedCompareExchangeRelease

Finally, the usage of these APIs at certain hot code paths (mostly in `syncblk.*` files) are replaced with the rewritten APIs.

This does introduce an extra check for the flag, but it is cheaper than any other alternative like function pointer. With function pointer, there has to be an extra call and it has to go through CFG which makes it more expensive.

Clang [added a flag](https://reviews.llvm.org/rG4d7df43ffdb460dddb2877a886f75f45c3fee188) `-moutline-atomics` recently that would emit such check during compilation so the faster machines can take advantage of it while still maintaining compatibility with the older machines. See an example [here](https://mysqlonarm.github.io/ARM-LSE-and-MySQL/). Unfortunately, we are still on clang9 and hopefully, this PR and the improvements I share below will motivate us to move to newer compiler on Linux to have this taken care by compiler. Hence, for now I just introduced this on Windows. The MSVC story on windows is slightly different. It doesn't have an equivalent flag like `-moutline-atomics` so the only option to generate faster atomic instructions would be to use `/arch:armv8.1` as mentioned [here](https://devblogs.microsoft.com/cppblog/msvc-backend-updates-in-visual-studio-2022-version-17-2/), which unfortunately we cannot do. So the choice was to add such check manually in our codebase to take advantage of the atomic instructions.

## Benchmark
I tested this PR with following C# code:

```c#
namespace Monitor_Lock  
{  
    class Program  
    {  
        static readonly object _object = new object();  
        static long result = 0;
  
        static void TestLock()  
        {
            for (int i = 0; i < 10000; i++)
            {
                lock (_object)  
                {  
                    result += i;
                }
            }
        }  
  
        static void Main(string[] args)      
        {
            Stopwatch sw = new Stopwatch();
            int count = int.Parse(args[0]);
            List<Thread> threads = new List<Thread>();
            sw.Start();
            for (int i = 0; i < count; i++)  
            {  
                ThreadStart start = new ThreadStart(TestLock);  
                Thread thread = new Thread(start);
                threads.Add(thread);
                thread.Start();
            }
            for (int i = 0; i < count; i++)
            {
                threads[i].Join();
            }
            sw.Stop();
            Console.WriteLine($"{count} threads took {sw.Elapsed.TotalMilliseconds} msec.");
        }
    }  
}  
```

## Results

### Ampere (v8.1 supported)

As part of the experiments, I ran this program on a 80-core Windows Ampere machine with varying number of thread counts and for each thread count, performed 5 iterations. Below is the AVG/MIN/MAX duration for running this test. 

Average duration:

| Threads AVG. | main      | PR        | Diff    |
|--------------|-----------|-----------|---------|
| 1            | 3.4344    | 3.4844    | 1.46%   |
| 2            | 4.47042   | 4.5869    | 2.61%   |
| 4            | 7.64708   | 6.0626    | -20.72% |
| 8            | 12.4943   | 9.6475    | -22.78% |
| 16           | 21.71328  | 20.94708  | -3.53%  |
| 32           | 42.86546  | 38.87922  | -9.30%  |
| 64           | 83.62778  | 70.58204  | -15.60% |
| 80           | 101.097   | 82.45124  | -18.44% |
| 128          | 167.65336 | 141.48888 | -15.61% |
| 256          | 316.17956 | 279.4261  | -11.62% |

Minimum duration:

| Threads MIN. | main     | PR       | Diff    |
|--------------|----------|----------|---------|
| 1            | 3.4012   | 3.4082   | 0.21%   |
| 2            | 4.3022   | 4.1128   | -4.40%  |
| 4            | 7.0673   | 5.7692   | -18.37% |
| 8            | 10.4781  | 8.4221   | -19.62% |
| 16           | 18.6477  | 16.4685  | -11.69% |
| 32           | 36.0364  | 33.6191  | -6.71%  |
| 64           | 71.2875  | 65.0999  | -8.68%  |
| 80           | 92.3571  | 72.0427  | -22.00% |
| 128          | 158.3358 | 134.0232 | -15.36% |
| 256          | 288.162  | 268.1084 | -6.96%  |

Maximum duration:

| Threads MAX. | main     | PR       | Diff    |
|--------------|----------|----------|---------|
| 1            | 3.4906   | 3.69     | 5.71%   |
| 2            | 4.6457   | 5.1342   | 10.52%  |
| 4            | 8.6484   | 6.6577   | -23.02% |
| 8            | 13.6105  | 11.0653  | -18.70% |
| 16           | 24.0912  | 24.7727  | 2.83%   |
| 32           | 49.7276  | 48.7453  | -1.98%  |
| 64           | 344.3766 | 298.0459 | -13.45% |
| 80           | 113.6214 | 89.741   | -21.02% |
| 128          | 178.9653 | 145.3084 | -18.81% |
| 256          | 344.3766 | 298.0459 | -13.45% |

Overall, I see generous improvements from 10-20% in synchronization methods. It is little puzzling why 16 threads would just show slight performance improvement and I will check with few people offline.

### Qualcomm (v8.1 not supported)

I also verified on 46-core Windows Qualcomm machine that doesn't have atomic instruction to make sure that we don't regress because of the extra flag checks. Here are the numbers (just posting the average)

| Threads | main (avg) | PR (avg) | diff   |
|---------|------------|----------|--------|
| 1       | 4.9163     | 4.97444  | 1.18%  |
| 2       | 7.31232    | 7.4033   | 1.24%  |
| 4       | 10.77444   | 11.34458 | 5.29%  |
| 8       | 17.89266   | 18.59982 | 3.95%  |
| 16      | 31.76536   | 31.9401  | 0.55%  |
| 32      | 59.63574   | 58.14956 | -2.49% |
| 46      | 82.8319    | 81.73412 | -1.33% |
| 64      | 115.20062  | 114.6887 | -0.44% |
| 128     | 223.50822  | 221.9457 | -0.70% |
| 256     | 448.14158  | 441.6222 | -1.45% |

Overall, the numbers are within error margin.


Here is how the code looks like (taken from windbg screenshots)

## In Action

Before:

![image](https://user-images.githubusercontent.com/12488060/174402366-40fd05c9-b947-41af-a490-22c291b151f0.png)

After:

![image](https://user-images.githubusercontent.com/12488060/174402372-c3a0b5f5-313e-49c1-8a6a-25d1ea9cbe35.png)

## Next steps
- If this is convincing enough, we should start the process to upgrade our Linux compiler from clang9 to clang12 or clang13.
- I will start adding few more APIs so we can optimize some of the hot code paths in GC.

kunalspathak committed:
Use Fast compareexchange, acquire/release

kunalspathak committed:
working windows version

kunalspathak committed:
remove printf

kunalspathak committed:
some more #ifdef

kunalspathak committed:
fix some #ifdef

kunalspathak committed:
optimize EnterObjMonitorHelperSpin

kunalspathak committed:
Remove FastInterlockedCompareExchange64 for now

jkotas commented:
> It is little puzzling why 16 threads would just show slight performance improvement and I will check with few people offline.

It may be caused by threads finishing the work before all threads are running. It is also why you do not see any improvement for 2 threads - there should be improvement for 2 threads too. You may want to change the spin count to `1_000_000` and add Barrier in frond the loop so that all threads wait for each other getting created before starting the work.

kunalspathak commented:
Thanks @jkotas . I updated to `100000` for shorter test run duration and also added memory barrier before starting the threads. Here is modified code along with test result. I think they all make sense now except the Min/2 threads, but if you see raw numbers, the variance for `main` is around 64%, so I won't trust that number.


```c#
namespace Monitor_Lock  
{  
    class Program  
    {  
        static readonly object _object = new object();  
        static long result = 0;
  
        static void TestLock()  
        {
            for (int i = 0; i < 100000; i++)
            {
                lock (_object)  
                {  
                    result += i;
                }
            }
        }  
  
        static void Main(string[] args)      
        {
            Stopwatch sw = new Stopwatch();
            int count = int.Parse(args[0]);
            List<Thread> threads = new List<Thread>();
            sw.Start();
            for (int i = 0; i < count; i++)  
            {  
                ThreadStart start = new ThreadStart(TestLock);  
                Thread thread = new Thread(start);
                threads.Add(thread);
            }

            Interlocked.MemoryBarrier();

            for (int i = 0; i < count; i++)  
            {  
                threads[i].Start();
            }

            for (int i = 0; i < count; i++)
            {
                threads[i].Join();
            }
            sw.Stop();
            Console.WriteLine($"{count} threads took {sw.Elapsed.TotalMilliseconds} msec.");
        }
    }  
}  
```
Average:

| Threads | main     | PR       | diff    |
|---------|----------|----------|---------|
| 1       | 16.74636 | 14.06296 | -16.02% |
| 2       | 18.3666  | 16.4355  | -10.51% |
| 4       | 43.69844 | 43.00226 | -1.59%  |
| 8       | 86.08646 | 77.678   | -9.77%  |
| 16      | 200.1058 | 167.4846 | -16.30% |
| 32      | 342.4692 | 308.6021 | -9.89%  |
| 64      | 605.3218 | 542.0234 | -10.46% |
| 80      | 807.5776 | 687.6071 | -14.86% |
| 128     | 1230.828 | 1167.448 | -5.15%  |
| 256     | 2674.654 | 2251.305 | -15.83% |

Min:
| Threads | main     | PR       | diff    |
|---------|----------|----------|---------|
| 1       | 5.1449   | 5.0627   | -1.60%  |
| 2       | 11.7118  | 15.5086  | 32.42%  |
| 4       | 36.8293  | 27.3163  | -25.83% |
| 8       | 74.6232  | 56.595   | -24.16% |
| 16      | 140.0395 | 132.3511 | -5.49%  |
| 32      | 325.5867 | 249.9301 | -23.24% |
| 64      | 578.1493 | 453.8749 | -21.50% |
| 80      | 700.3981 | 562.5034 | -19.69% |
| 128     | 1143.339 | 1060.977 | -7.20%  |
| 256     | 2413.856 | 2139.666 | -11.36% |

Max:

| Threads | main     | PR       | diff    |
|---------|----------|----------|---------|
| 1       | 5.6196   | 5.6733   | 0.96%   |
| 2       | 31.1041  | 17.2587  | -44.51% |
| 4       | 61.231   | 61.2019  | -0.05%  |
| 8       | 102.0049 | 100.7827 | -1.20%  |
| 16      | 245.4816 | 204.5831 | -16.66% |
| 32      | 354.2334 | 359.8534 | 1.59%   |
| 64      | 628.9136 | 601.3752 | -4.38%  |
| 80      | 1041.367 | 799.7853 | -23.20% |
| 128     | 1336.44  | 1259.636 | -5.75%  |
| 256     | 2803.754 | 2360.407 | -15.81% |

<details>
<summary> Raw numbers of all iterations </summary>

| Threads | main     | PR       |
|---------|----------|----------|
| 1       | 5.6196   | 5.6733   |
| 1       | 5.1449   | 5.0627   |
| 1       | 5.1681   | 5.1654   |
| 1       | 5.1943   | 5.1351   |
| 1       | 5.3091   | 5.2476   |
| 2       | 13.2151  | 16.3577  |
| 2       | 11.7118  | 15.9422  |
| 2       | 22.3917  | 17.2587  |
| 2       | 31.1041  | 15.5086  |
| 2       | 13.4103  | 17.1103  |
| 4       | 43.1938  | 35.0976  |
| 4       | 37.7045  | 52.4248  |
| 4       | 39.5336  | 61.2019  |
| 4       | 61.231   | 27.3163  |
| 4       | 36.8293  | 38.9707  |
| 8       | 79.4986  | 56.595   |
| 8       | 74.6232  | 69.2368  |
| 8       | 78.4101  | 83.1975  |
| 8       | 102.0049 | 100.7827 |
| 8       | 95.8955  | 78.578   |
| 16      | 241.5675 | 193.9641 |
| 16      | 148.6181 | 132.3511 |
| 16      | 224.8223 | 170.8608 |
| 16      | 245.4816 | 135.664  |
| 16      | 140.0395 | 204.5831 |
| 32      | 341.4042 | 249.9301 |
| 32      | 325.5867 | 343.8633 |
| 32      | 338.0256 | 359.8534 |
| 32      | 354.2334 | 315.5561 |
| 32      | 353.0962 | 273.8075 |
| 64      | 628.9136 | 569.8978 |
| 64      | 578.1493 | 566.4561 |
| 64      | 587.394  | 453.8749 |
| 64      | 627.6597 | 601.3752 |
| 64      | 604.4926 | 518.5128 |
| 80      | 831.2326 | 571.1818 |
| 80      | 700.3981 | 755.7222 |
| 80      | 1041.367 | 562.5034 |
| 80      | 732.4616 | 799.7853 |
| 80      | 732.4285 | 748.8428 |
| 128     | 1336.44  | 1232.302 |
| 128     | 1219.41  | 1060.977 |
| 128     | 1174.404 | 1259.636 |
| 128     | 1280.545 | 1181.896 |
| 128     | 1143.339 | 1102.428 |
| 256     | 2775.944 | 2139.666 |
| 256     | 2413.856 | 2265.159 |
| 256     | 2703.091 | 2194.7   |
| 256     | 2803.754 | 2296.592 |
| 256     | 2676.625 | 2360.407 |
</details>

jkotas commented:
>  also added memory barrier before starting the threads.

Nit: I meant `Barrier` not memory barrier. Something like:
```
using System.Threading;
using System.Diagnostics;

class Program
{
    static readonly object _object = new object();
    static long result = 0;
    static int _iterations;
    static Barrier _barrier;

    static void TestLock()
    {
        for (;;)
        {
            _barrier.SignalAndWait();
            int iterations = _iterations;
            for (int i = 0; i < iterations; i++)
            {
                lock (_object)
                {
                    result += i;
                }
            }
            _barrier.SignalAndWait();
        }
    }

    static void Main(string[] args)
    {
        Stopwatch sw = new Stopwatch();
        int count = int.Parse(args[0]);
        _iterations = 10_000_000 / count;
        _barrier = new Barrier(count + 1);
        for (int i = 0; i < count; i++)
        {
            new Thread(TestLock) { IsBackground = true }.Start();
        }

        for (int i = 0; i < 5; i++)
        {
            _barrier.SignalAndWait();
            sw.Restart();
            _barrier.SignalAndWait();
            if (i != 0) // Throw away warm up run
               Console.WriteLine($"{count} threads took {sw.Elapsed.TotalMilliseconds} msec.");
        }
    }
}
```

kunalspathak commented:
> Nit: I meant `Barrier` not memory barrier. Something like:

I saw some variance so I reran the tests multiple times to get this data.

| Threads | main     | PR       | diff    |
|---------|----------|----------|---------|
| 1       | 263.7254 | 254.9208 | -3.34%  |
| 2       | 1221.154 | 929.6501 | -23.87% |
| 4       | 1119.331 | 841.3092 | -24.84% |
| 8       | 1226.24  | 1019.737 | -16.84% |
| 16      | 1326.107 | 1135.433 | -14.38% |
| 32      | 1298.534 | 1163.145 | -10.43% |
| 64      | 1468.626 | 1129.997 | -23.06% |
| 80      | 1455.727 | 1089.452 | -25.16% |
| 128     | 957.8012 | 809.328  | -15.50% |
| 256     | 1220.782 | 767.7598 | -37.11% |

<details>
<summary>Raw numbers in msecs</summary>

| iter | threads | main     | PR       |
|------|---------|----------|----------|
| 1    | 1       | 264.6954 | 255.9875 |
| 2    | 1       | 263.549  | 254.9277 |
| 3    | 1       | 263.4125 | 254.3808 |
| 4    | 1       | 263.5136 | 254.571  |
| 5    | 1       | 263.4564 | 254.7372 |
| 1    | 2       | 1105.415 | 1063.21  |
| 2    | 2       | 1371.401 | 906.5986 |
| 3    | 2       | 997.5693 | 916.9482 |
| 4    | 2       | 1279.715 | 1022.461 |
| 5    | 2       | 1351.671 | 739.0335 |
| 1    | 4       | 1139.591 | 847.6639 |
| 2    | 4       | 1115.59  | 866.1285 |
| 3    | 4       | 1119.016 | 861.1583 |
| 4    | 4       | 1123.302 | 848.6115 |
| 5    | 4       | 1099.155 | 782.984  |
| 1    | 8       | 1397.861 | 1286.234 |
| 2    | 8       | 1130.025 | 1270.928 |
| 3    | 8       | 1112.098 | 886.4905 |
| 4    | 8       | 1310.242 | 818.4406 |
| 5    | 8       | 1180.976 | 836.5925 |
| 1    | 16      | 1319.007 | 1135.006 |
| 2    | 16      | 1381.922 | 1124.878 |
| 3    | 16      | 1387.507 | 1124.588 |
| 4    | 16      | 1314.842 | 1142.788 |
| 5    | 16      | 1227.254 | 1149.904 |
| 1    | 32      | 1268.167 | 1064.065 |
| 2    | 32      | 1358.571 | 1218.19  |
| 3    | 32      | 1296.766 | 1176.425 |
| 4    | 32      | 1279.84  | 1133.739 |
| 5    | 32      | 1289.324 | 1223.305 |
| 1    | 64      | 1516.131 | 1181.214 |
| 2    | 64      | 1462.303 | 1116.237 |
| 3    | 64      | 1461.59  | 1094.956 |
| 4    | 64      | 1483.634 | 1184.118 |
| 5    | 64      | 1419.474 | 1073.458 |
| 1    | 80      | 1492.395 | 1082.396 |
| 2    | 80      | 1442.627 | 1099.673 |
| 3    | 80      | 1464.661 | 1071.737 |
| 4    | 80      | 1400.441 | 1105.974 |
| 5    | 80      | 1478.513 | 1087.477 |
| 1    | 128     | 943.9751 | 814.531  |
| 2    | 128     | 970.4388 | 821.8595 |
| 3    | 128     | 986.5561 | 805.6217 |
| 4    | 128     | 943.8825 | 806.1945 |
| 5    | 128     | 944.1534 | 798.4335 |
| 1    | 256     | 1237.129 | 757.3509 |
| 2    | 256     | 1250.24  | 786.2865 |
| 3    | 256     | 1239.037 | 756.1162 |
| 4    | 256     | 1175.578 | 775.7833 |
| 5    | 256     | 1201.925 | 763.2619 |

</details>

kunalspathak commented:
@davidwrighton , @Maoni0 , @janvorli , @dotnet/jit-contrib

EgorBo commented:
I assume it closes https://github.com/dotnet/runtime/issues/67824

kunalspathak commented on src/coreclr/vm/util.hpp:
```diff
@@ -71,6 +79,64 @@ BOOL inline FitsInU4(unsigned __int64 val)
+{
+    if (g_arm64_atomics_present)
+    {
+        return (LONG) __casal32((unsigned __int32*) Destination, (unsigned  __int32)Comperand, (unsigned __int32)Exchange);
```
`__casal32` is an intrinsic exposed by VC++ only. For osx-arm64, we can just patch `PAL_ArmInterlockedOperationBarrier` to not call `__sync_synchronize`, right?

EgorBo replied:
> For osx-arm64, we can just patch PAL_ArmInterlockedOperationBarrier to not call __sync_synchronize, right?

Ah, no, that will depend on compilers flag, e.g. if someone adds `-mcpu=generic` to CMake then clang will stop emitting e.g. `casal` for us on osx-arm64 and the barrier will be needed. While an intrinsic would always be fine since there is no pre 8.1 hw on osx-arm64 (and will never be judging by apple's clang)

kunalspathak replied:
> if someone adds `-mcpu=generic` to CMake

What is the default flag used for osx? Since its basline is 8.1, it should be `-mcpu=armv8.1` by default, even if we don't specify it? 

> While an intrinsic would always be fine since there is no pre 8.1 hw on osx-arm64

I don't understand how/why will someone want to use  `-mcpu=generic` and even if they use it, it will be `v8.1` only. Correct me if I am wrong.

> While an intrinsic would always be fine 

Can you check what code we generate today with default flags on osx for `__sync_val_compare_and_swap`? Should be `casal`. In that case, we can eliminate the `__sync_synchronize` for OSX. What change you tried that you mentioned in https://github.com/dotnet/runtime/issues/67824#issuecomment-1094337248?

EgorBo replied:
> What change you tried that you mentioned in https://github.com/dotnet/runtime/issues/67824#issuecomment-1094337248?

I think it was `-mcpu=native` (I verified coreclr via objdump) without that explicit barrier for all Interlocked functions including the one is GC and measured impact on TE:
![image](https://user-images.githubusercontent.com/523221/174631146-ee3eedf1-ccd7-4fa2-b0b7-44119c1d3535.png)

> What is the default flag used for osx? 

We don't set march/mcpu/m* flags so we use defaults

kunalspathak replied:
- Use `-mcpu=native`. I don't think there is any disadvantage with this and in fact, Arm documentation [prefers to have that](https://community.arm.com/arm-community-blogs/b/tools-software-ides-blog/posts/compiler-flags-across-architectures-march-mtune-and-mcpu).  Given that it was [added in 2018](https://github.com/llvm-mirror/clang/commit/86c991513001535af6b82bcb1f7c45ab60d2adf0), it should be available in clang9, but its availability is questionable for osx, see [this](https://stackoverflow.com/questions/65966969/why-does-march-native-not-work-on-apple-m1) and [this](https://github.com/UZ-SLAMLab/ORB_SLAM3/issues/542).
- For osx, we should definitely remove the barriers from `pal.h` because the baseline for M1 is `v8.1+` AFAIK. @VSadov - any objections?

VSadov replied:
> For osx, we should definitely remove the barriers from pal.h because the baseline for M1 is v8.1+ AFAIK.

The barrier is a workaround for `__sync_val_compare_and_swap` not being a full fence. 

M1 is v8.1+, so `casal` would be sufficient and additional fence is redundant. 
but does compiler actually use `casal`?

kunalspathak replied:
> but does compiler actually use casal

It does. https://godbolt.org/z/8xb3x7o71

I will put a PR out.

EgorBo replied:
if `-mcpu=generic` somehow flows in it will stop emitting casal

kunalspathak replied:
https://github.com/dotnet/runtime/pull/71026

VSadov replied:
Does it use `casal` in debug? Can it switch to old helper because of register pressure or if the old implementation is found faster (it could be).  
It feels like inline asm could have more reliable guarantees.

kunalspathak commented:
> Do you plan to work on upgrading the compiler so that we can take advantage of this?

I chatted with @davidwrighton  and we will work with @janvorli to do this.

VSadov commented on src/coreclr/vm/syncblk.inl:
```diff
@@ -723,7 +731,12 @@ FORCEINLINE AwareLock::LeaveHelperAction ObjHeader::LeaveObjMonitorHelper(Thread
             // We are leaving the lock
             DWORD newValue = (syncBlockValue & (~SBLK_MASK_LOCK_THREADID));
+
+#if defined(TARGET_WINDOWS) && defined(TARGET_ARM64)
```
I think moving the static/dynamic switching logic into `InterlockedCompareExchangeRelease` and similar would be nicer.

It should result in simpler looking code and if switching needs to change due to new compiler or platform, no need to revisit all the callsites.

VSadov commented on src/coreclr/vm/syncblk.inl:
```diff
@@ -650,7 +654,11 @@ FORCEINLINE AwareLock::EnterHelperResult ObjHeader::EnterObjMonitorHelper(Thread
     }
 
+#if defined(TARGET_WINDOWS) && defined(TARGET_ARM64)
+    if (FastInterlockedCompareExchangeAcquire((LONG*)&m_SyncBlockValue, newValue, oldValue) == oldValue)
```
Didn't we just remove some or all of the "FastInterlockedXXXX" versions?

Having "Fast" variants always feels a bit confusing. Are they unfit for some uses? Where would want to be using the slow versions?

kunalspathak replied:
Yes, we wanted to just patch the hot paths (which is where we should always use fast versions) instead of all the paths and hence I had to reintroduce them at just few call-sites.

kunalspathak commented:
Improvements: https://github.com/dotnet/perf-autofiling-issues/issues/6347
